coding_standard_c.txt

```diff
@@ -0,0 +1,147 @@
+ */
+/* standard #includes */
+/* project #includes */
+#include “basic_file.h”
+/* externalized variable definitions */
+/* static variables (locally global) */
+/* static function declarations */
+/* externalized function definitions */
+/* static function definitions */
+
+Header files should only #include what is necessary to allow a file that includes it to compile.  Associated source files will always #include the header of the same name, but should #include files whose resources are used within the source even if they are already included in that header.  This provides a complete context for readers of the source file... i.e., they don't have to search through headers to determine  where a resource came from.  For example, consider a header and source that use FILE pointers declared in stdio.h.  The header must #include <stdio.h> to ensure files that include it will be able to compile.  The corresponding source #includes the associated header and therefore receives the benefit of stdio.h automatically.  The source would be unnerving if the FILE type was used and stdio.h was not #included.
+
+Variable Scope and types
+Variables should be declared and scoped to the smallest scope of use.
+
+/* This should be avoided */
+void foo() {
+	unsigned i;
+	do_something();
+	for(i=0; i < 100; i++) {
+		do_more(i);
+	}
+}
+
+/* Proper variable scope */
+void foo() {
+
+	do_something();
+	unsigned i;
+	for(i=0; i < 100; i++) {
+		do_more(i);
+	}
+}
+
+Variables should only be of a signed type if something should ever be negative. A common, incorrect use, is to declare loop indices as int instead of unsigned.
+
+Formatting
+Always use tab characters rather than spaces at the beginning of lines.  Make sure your editing environment uses at least five visual spaces to represent tabs.  This will make code push horizontal limits with very little depth, thus deterring extraordinary amounts of nested statements.  The large tab approach “makes poorly-written programs blatantly ugly”. [2]  It may periodically cause small pieces of well written code to seem unruly, but it is worth it overall.
+
+Block statements encapsulated with braces should place the opening brace on the same line after the end of the conditional statement or function parameter list, and the closing brace at the same column position as the beginning of the conditional statement or function return parameter, on the line immediately after the end of the block.
+
+if(some_int > 0) {
+     statement1;
+     statement2;
+}
+
+void some_function(arg1, arg2) {
+     statement1;
+     statement2;
+}
+
+Conditional statements (such as if, else, while, for, switch, etc) should always be followed by such block statements, even if they contain only one line statement.  These formatting conditions are contrary to Kernighan and Ritchie's “one true brace style”. [3]
+
+Line Length should not exceed 100 characters and should be split on the nearest whitespace or delimiter character.
+
```

Comment:
Is 80 characters too limited? I started sticking to 80 just on principle but found that much more than 80 was a strain. Scanning my eyes horizontal too far causes me to lose track of where I started. This isn't a sticking point for me though since we've got source here that wraps across my screen and getting down under 100 would be an improvement.

Also there are some lines in this document well over 100 characters long :smile: